### PR TITLE
Diffing of blocks from decompiled TS

### PR DESCRIPTION
**Pull request by pelikhan**
Uses the TS diff to render a blocks diff. This allows to compute diffs between documentation snippets and tutorial steps; or between blocks that have been rountripped to JS.

* run http://localhost:3232/release-tests/md/diff for testing
* guarded by pxtarget flag to experiment
* automatic step diffing in blocks tutorials

* rendering of diffs in docs

- [x] generate source maps on decompilation
- [x] added 'diffblocks' section
- [x] render js/py diff as well
- [x] handle trailing ; in sources
- [x] auto-diffing in tutorials
- [x] minimize size of serialized tutorial data
- [x] added test + updated test suite (marked as noDiffs)

### CODE REVIEW

File: localtypings/pxtarget.d.ts [hunk 1]
```diff
@@ -855,7 +855,6 @@ declare namespace pxt.tutorial {
         // no coding
         unplugged?: boolean;
         tutorialCompleted?: boolean;
-        hasHint?: boolean;
```
Comment (pelikhan) [inline localtypings/pxtarget.d.ts]: inferred from ``!!hintContentMd``

File: pxtlib/tutorial.ts [hunk 2]
```diff
@@ -168,15 +195,19 @@ ${code}
...
-                unplugged: /@unplugged/.test(flags),
-                tutorialCompleted: /@tutorialCompleted/.test(flags),
-                resetDiff: /@resetDiff/.test(flags),
                 contentMd: step,
-                headerContentMd: header,
-                hintContentMd: hint,
-                hasHint: hint && hint.length > 0
+                headerContentMd: header
             }
+            if(/@(fullscreen|unplugged)/.test(flags))
```
Comment (pelikhan) [inline pxtlib/tutorial.ts]: avoid useless fields in serialized json

**Comment by abchatra** [comment]
Can we just highlight the block diffed by color and keep the rest the same color? Gray color is hard to read in my opinion.

**Comment by pelikhan** [comment]
The whole point is to focus the attention on the changes on the changes blocks.

**Comment by abchatra** [comment]
Yes highlight it without making everything else look disabled. Like in the 'python' yellow highlighting.

**Comment by pelikhan** [comment]
The highlight does not focus the attention enough , imo.

In python the text is bw by default, which is quite different from the blocs filled with colors.

**Comment by pelikhan** [comment]
This is a UI design question.

**Comment by abchatra** [comment]
> This is a UI design question.

Yes, we can discuss tomorrow in UI meeting.

**Comment by darzu** [comment]
Love it. I like it how it's designed right now. Everything else grey and new thing is colored.

File: pxtblocks/blocklydiff.ts [hunk 3]
```diff
@@ -401,4 +422,75 @@ namespace pxt.blocks {
...
+                    }
+                    // find block ids mapped to the ranges
+                    const oldid = pxt.blocks.findBlockId(oldResp.sourceMap, {
+                        start: oldLineStart,
+                        length: lineLength
+                    });
+                    const newid = pxt.blocks.findBlockId(newResp.sourceMap, {
+                        start: newLineStart,
+                        length: lineLength
+                    });
```
Comment (darzu) [inline pxtblocks/blocklydiff.ts]: After my PR, you can provide a position within a single line, in case that helps

Comment (pelikhan) [reply pxtblocks/blocklydiff.ts]: right now i don't handle line diff but eventually we could be more precise and flag changes within nested blocs

Comment (pelikhan) [reply pxtblocks/blocklydiff.ts]: or after my PR, you can update my decompiler. why do we need this?

**Comment by abchatra** [comment]
Can we add flag to enable\disable this feature in the target on need basis? We have a deadline for Minecraft and hour of code is dependent on the hints. Also hour of code is just one step.

File: pxtblocks/blocklydiff.ts (hunk 3 above)
Comment (darzu) [reply pxtblocks/blocklydiff.ts]: I think the use case is for tutorial steps that just modify a line like "change this number" or "add (pick random) into the second argument". Not sure how many non-trivial examples of this we have.

**Comment by pelikhan** [comment]
You can also disable the feature for a tutorial using
```
### @noDiffs
```

**Pull request merged**